spec/support/change_streams.rb

```diff
@@ -0,0 +1,266 @@
+          @pipeline = test['changeStreamPipeline'] || []
+          @options = test['changeStreamOptions'] || {}
+          @operations = test['operations'].map { |op| Operation.new(op) }
+          @expectations = test['expectations']
+          @result = test['result']
+          @coll1_name = coll1
+          @coll2_name = coll2
+          @db1_name = db1
+          @db2_name = db2
+        end
+
+        def setup_test
+          @global_client = ADMIN_AUTHORIZED_TEST_CLIENT.use('admin')
+
+          @db1 = @global_client.use(@db1_name).database.tap(&:drop)
+          @db2 = @global_client.use(@db2_name).database.tap(&:drop)
+
+          @db1[@coll1_name].create
+          @db2[@coll2_name].create
+
+          client = ADMIN_AUTHORIZED_TEST_CLIENT.with(
+            database: @db1_name,
+            app_name: 'this is used solely to force the new client to create its own cluster')
+          client.subscribe(Mongo::Monitoring::COMMAND, EventSubscriber.clear_events!)
+
+          @target = case @target_type
+                   when 'client'
+                     client
+                   when 'database'
+                     client.database
+                   when 'collection'
+                     client[@coll1_name]
+                    end
+        end
+
+        def run
+          change_stream = begin
+            @target.watch(@pipeline, @options)
+          rescue Mongo::Error::OperationFailure => e
+            return {
+              result: { 'error' => { 'code' => e.code } },
+              events: events
+            }
+          end
+
+          @operations.each do |op|
+            op.execute(@db1, @db2)
+          end
+
+          changes = [].tap do |changes|
+                           next unless @result['success']
+
+                           unless @result['success'].empty?
+                             change_stream.take_while do |change|
+                               changes << change
+                               changes.length < @result['success'].length
+                             end
+                           end
+
+                           change_stream.close
+                         end
+
+          {
+            result: { 'success' => changes },
+            events: events,
+          }
+        end
+
+        def teardown_test
+          @global_client.close
+          @db1.drop
+          @db2.drop
+        end
+
+        def match_result?(result)
+          case @result.first.first
+          when 'success'
+            match_success?(result)
+          when 'error'
+            @result == result
+          end
+        end
+
+        def match_commands?(actual)
+          @expectations.each_with_index.all? do |e, i|
+            actual[i] && match?(e, actual[i])
+          end
+        end
+
+        private
+
+        def events
+          EventSubscriber.started_events.reduce([]) do |evs, e|
+            next evs if [:saslStart, :saslContinue, :killCursors].include?(e.command_name)
```

Comment:
command_name is now always a string thus this shouldn't work.

Reply:
Interestingly, it does work though; I inspected the output with and without this line, and the sasl and killCursor commands are indeed removed (and the branch is rebased against master).

Reply:
Okay great.